### DISCUSSION TITLE
F2F: Class initialization in qbicc and Leyden

### DISCUSSION THREAD

**Discussion by dmlloyd** [qbicc/qbicc#523] [Design discussions]
# Links

* The [recording](https://drive.google.com/file/d/1afUEJ7MJg3ozHQ9eSHisiamugR7AsbBD/view)

# Attendees

* @ashu-mehra 
* @dgrove-oss 
* @DanHeidinga 
* @kazunoriogata 
* @theresa-m 
* @evacchi 
* @n1hility 
* @dmlloyd 

# Agenda

* :00 Introduction and overview
* :05 @dgrove-oss - overview of heap serialization mechanism
* :15 @dmlloyd - constant initialization is currently broken 😞
* :20 @DanHeidinga - overview of initialization work to date and research results
* (transition to) open discussion
* 1:25 - wrap up

# Notes

* Heap serialization
   * Build heap at build time, then get this into the image at runtime
   * Two options: 
      * save heap and restore 
      * Binary serialization stream (our approach)
   * Register static field, here’s the object graph to be reachable from the heap, and that lets us know the types that need special deserializers
   * Basic serialization / deserialization working
   * TODO: change to writing target endian
      * CompilationContext probe() should be able to provide this info
      * Some work required to get streams that write in LE
* Constant initialization
   * Previously was single threaded so initializers were mostly enqueued and processed immediately after having been seen.
   * Parallel compilation means initializers aren’t parsed until later, so somewhat random on when initializers get parsed and therefore constant folded
   * Goal - want the initializers parsed first so we can do direct substitution in the rest of the graph
   * May need two phases, one to find initiailizers and one to pass through
   * Or use a lock and follow the JVMS on processing initializers as they get encountered
   * Constant replacement is a bit hacky - two passes one in add and one in analyze
   * Ideally, BTI the classes when encountering initializers
   * Without frontloading the constants, we’d have an explosion of code to parse (ie platform specific code)
      * Can we special case some of these constants?
      * One option to use an annotation to mark fields needing aggressive processing during class load
      * Analogous to bootstrapping problems
      * Remember one of the goals - can we do this quickly but llvm execution dominates
   * Similarly to graal, need to be able to handle the equiv of #ifdef in our java transliterated C
   * Figure out which ~20 classes need to be done first and do those before doing main
* Class Initializers
   * preserve semantics (e.g. Random initializers, capturing environment, e.g. n. CPUs) 
   * How may devs express intent?
   * no way to indicate "at what time" to perform init
   * We don't want to "break" the platform, not throw away the dynamic JVM
   * Couple of options for build-time init
      * Splitting class init into 2 methods build-time + run-time
      * Modify bytecode?
      * No build-time init at all, just record all initializers, and evaluate at immediate startup, first thing
   * not initializers, but "re-initializers" ? when you "restart"
   * GraalVM's Substitutions -- mental model: class file load hook
   * Caveat with spec: failed initializers should report errors "as if" were initialized at that time (regardless of when the init code has run).
   * Re-init: e.g. random number initialization, you don't want the build-time seed to recycle at run-time
   * Spec: assumption other JDKs can implement the requirements for native image
   * Substitutions would never be accepted in the spec -- could be a separate side-project?
   * Substitution is really a class-loading hook, i.e. a JDK agent?
* Benefits of BTI
   * What do we want BTI for?
   * we get a ready-to-use heap
   * => fast startup, being able to scale instantly (e.g. serverless)
   * map heap instead of load and fix it up (snapshot in GraalVM)
      * Eduardo: To make this as fast as possible, the heap start is always available in a fixed register (we use the register r14 on x64 architectures).
* Interpreter
   * bytecode interpreter with C.Nutter (JRuby)
   * instead execute the program graph directly
* Other ways to get an initial heap?
   * "fun" data point. A basic Java hello world requires ~120 classes
   * class initializers is a hook that can do anything! 
   * Retrospectively, is this a "good" mechanism for Leyden at all?
   * Constant Dynamic independent from class init
   * perspective of being lazy
   * "before" is a very big time -- for Leyden it could be build-time
   * you can't expect libraries to be Leyden ready before Leyden
   * engineering perspective: no boot-time heap, default to run-time then introduce BTI incrementally?
   * the point of this effort is to get boot-time heap, otherwise no real benefits wrt AppCDS
* Interpreter
   * proposal: try to execute and serialize; throw if it can't eval => exec at runtime
   * Analysis of the initializers to decide if they are safe to run?
* Can we rely on the annotations that are used in Quarkus?
   * provide keywords in the JVMLS ?
* static initializers are used to build structures, in Quarkus code is in static initializers so that startup is quick  
   * Including reading files at buildtime
* Use of Graal’s “this is buildtime” test to differentiate between buildtime and runtime init patterns
* Can we detect which code should be run at BT with static analysis?
   * Issue of intent and innocuous changes forcing related classes to be RTI
   * Intent sometimes can’t be figured out ie: reading a file, or getting a timestamp representing build time

## Next steps
* Fix the constants issue (make non-final, or annotation): David
* Proceed with RTI of everything: Dan
* Focus on figuring out what should be BTI vs RTI
   * What can we do with existing classfiles?
* Interpreter - try and BTI but throw on “bad” patterns and then defer to the RTI
* Prototype patches to classfiles to test expressing user intent

# Background reading

* "[Initialize Once, Start Fast: Application Initialization at Build Time](http://www.christianwimmer.at/Publications/Wimmer19a/Wimmer19a.pdf)" Wimmer et al.

**Comment by dgrove-oss** [comment]
Here's some pre-meeting notes on heap serialization mechanism.

## Overview

One step in producing a native executable for an application is evaluating selected application and library static initializers at compile time and embedding the resulting Java heap objects into the produced executable.

Our strategy for doing this is to build up the relevant Java heap in the compiler's heap, using the host JVM's heap.  We then traverse this object graph using reflection and serialize it to a compact binary representation that is emitted into the generated .ll files. The compiler also generates class specific deserialization functions into the same .ll file that will be used at runtime. During an "early" runtime phase, this binary representation is processed to create an initial Java heap for the application. 

An alternate approach, which we decided not to pursue,  is emitting a byte[] containing a fully formed initial heap into a .ll file. This avoids the serialization/deserialization steps, but either (a) requires assuming a known heap start address or (b) performing a pass through the heap at runtime to adjust internal heap pointers once the  start address is known. It also forces all initial heap objects into a single, non-GC managed memory space.

## Implementation

The compile-time side is in [org.qbicc.plugin.serialization](https://github.com/qbicc/qbicc/tree/main/plugins/serialization).

The runtime side is in [org.qbicc.runtime.deserialization](https://github.com/qbicc/qbicc/tree/main/runtime/deserialization).

Key Concepts:
* We support circular data structures by maintaining an IdentityHashMap of previously serialized objects and emitting a special "backref" tag on subsequent serialization attempts.
* The binary format attempts to efficiently encode common situations (eg. null, small arrays, small strings, "close" back references).
* The compiler generates class-specific deserialization methods for exactly those classes that had serialized instances to minimize meta-data in the binary stream.

What is implemented is pretty close to the initial design described in detail in https://github.com/qbicc/qbicc/pull/309, with some modifications to the "tagging scheme" to save some bytes here and there.

## Current Status

There is a functional implementation, but it has some limitations.

The process is driven by a `BuildtimeHeap` which defines the roots of the heap to be serialized.  It currently assumes each root is a static field (FieldElement).  We'll probably want to extend this to include at least global symbols and perhaps additional LLVM-level variables.

Each heap root is assumed to be class instance (plus its reachable heap). There is no support for serialization top-level primitive data (the assumption was this could be handled by emitting an initialized primitive value instead of using serialization).

Instances of java.lang.Class are not currently serialized (as we have not defined a runtime representation for java.lang.Class objects).

Class instances with non-trivial native resources "hidden" in primitive fields (threads, mutexes, file descriptors, etc) will be naively serialized by just writing their Java-level values (int, long, etc.). This is almost always doomed to fail because the matching native resources will not exist at runtime.

We currently serialize in big-endian format.  Once we have our build-time constant support working, we should instead serialize using target platform endianness. Using target platform endianness, will enable using bulk memory copy operations for deserializing primitive arrays.

**Comment by DanHeidinga** [comment]
And some pre-meeting notes on Class initialization.

## Overview

Broadly, there are two times at which a class can be initialized - either at buildtime (BTI) or runtime (RTI).  A dynamic JVM only allows RTI today and follows the processes defined in the JVM spec to detect when a class, and its supers, must be initialized.  A native image can move some, or possibly even all, class initialization to build time.

How to successfully employ BTI, in a way that respects the programmers intent, is an open question.  SubstrateVM demonstrates one set of options and they have highlighted some of the challenges for users with BTI in https://github.com/vjovanov/taming-build-time-initalization.

## Challenge

The challenge native images face is finding the right model for BTI that provides the most benefits (smaller image sizes and faster startup) without jettisoning the behaviour of the dynamic JVM.  Our task is to explore this space in a way that can inform the efforts for OpenJDK's Project Leyden.

We focus on BTI while recognizing that there will always be classes that need to be initialized - either partly or completely -  at runtime.  This covers cases like Random number generators, initializers that call into JNI, or take other actions that need to be dependent on the runtime environment rather than the build environment.

The work I've been doing so far has been adding support for class initialization that is conservatively correct with respect to the JVMS.  This allows for RTI with an eventual focus on "turning the dial" as far to BTI as possible.

See #530 and #458 (now #532)

## Principles

Find solutions to specifying BTI that respect the RTI environment so that there isn't a divergence between the dynamic JVM and native images.  While we want the benefits of native image, we don't want to split the platform or break compatibility with the dynamic JVM.

Prefer solutions that let the programmer specify their intent in the source code.  Respect the author's intent.  While exploring the technical solutions to enabling BTI, we should also be thinking about the user model for how they will tell the runtime (dynamic or static) when they want their initialization to occur.  This is similar to the approach taken with interface default methods where the author of the API can add them, but others outside the interface cannot extend it with their own default methods.

## Options

Some of the options to discuss and explore follow.  These are intended to jump start the conversation:

* Splitting current `<clinit>` methods into two `<static-clinit>` and `<clinit>` with a standardized approach for when and how to run them in dynamic and static models.

* Field annotations to indicate which fields should be BTI and which should be RTI.

* Run all the `<clinit>` methods that cannot be run at BT eager when starting the image.

* Explore a "Lifecycle API" which specifies how classes should be re-initialized or patched when a native image starts.  There's potentially a common approach here between native images and CRIU-style snapshots.

* Graal-style Substitutions to address modifying `<clinit>`s to be BTI ammenable
